### PR TITLE
CLI Application to print list of Objects given by a particular type.

### LINKED ISSUE #6: Print the list of objects from the spec identified by the type specified as CLI parameter
Parse an OpenAPI spec(input: URL) and print the list of objects from the spec identified by the type specified as CLI parameter.

The possible types are listed [here](https://spec.openapis.org/oas/latest.html#fixed-fields-5)
schemas
responses
parameters
examples
requestBodies
headers
securitySchemes
links
callbacks
pathItems

This is a qualifying task for GSoC'23.

### PR DESCRIPTION
**Pull request by Polaris66**
CLI Application that prints all Objects given by a particular type.

Resolves #6

Also, accesible through npm package - openapi-type-finder.

Application has been tested succesfully for all required types-
schemas
responses
parameters
examples
requestBodies
headers
securitySchemes
links
callbacks
pathItems

Even though all types have been seperately tested, there is a chance of errors. If such a thing happens please inform me.

I hope this solution meets the requirements for the Qualification Task.
Please give feedback however and whenever possible.

### CODE REVIEW

File: package/index.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
+#!/usr/bin/env node
+
+require("util").inspect.defaultOptions.depth=null
```
Comment (MikeRalphson) [inline package/index.js:3]: You learn something new every day! Thank you.

**Review by MikeRalphson — COMMENTED** [review]
Looks great - documentation is good. For a production tool, would need some live tests, but not needed here.

**Comment by bavulapati** [comment]
@Polaris66 Can you provide an example command to test the npm package?

**Comment by bavulapati** [comment]
> @Polaris66 Can you provide an example command to test the npm package?

Oh! I can see them on the npmjs.

**Comment by bavulapati** [comment]
@Polaris66 I'm running into an error, while using the npm package 

```
> ./node_modules/.bin/openapi-type-finder -u https://api.apis.guru/v2/specs/xkcd.com/1.0.0/openapi.yaml -t responses
node:internal/modules/cjs/loader:1093
  throw err;
  ^

Error: Cannot find module 'express/lib/response'
Require stack:
- /Users/balakrishna/projects/postman-open-technologies/openapi-componentizer/polaris/node_modules/openapi-type-finder/document.js
- /Users/balakrishna/projects/postman-open-technologies/openapi-componentizer/polaris/node_modules/openapi-type-finder/index.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:1090:15)
    at Module._load (node:internal/modules/cjs/loader:934:27)
    at Module.require (node:internal/modules/cjs/loader:1157:19)
    at require (node:internal/modules/helpers:119:18)
    at Object.<anonymous> (/Users/balakrishna/projects/postman-open-technologies/openapi-componentizer/polaris/node_modules/openapi-type-finder/document.js:4:18)
    at Module._compile (node:internal/modules/cjs/loader:1275:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1329:10)
    at Module.load (node:internal/modules/cjs/loader:1133:32)
    at Module._load (node:internal/modules/cjs/loader:972:12)
    at Module.require (node:internal/modules/cjs/loader:1157:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/Users/balakrishna/projects/postman-open-technologies/openapi-componentizer/polaris/node_modules/openapi-type-finder/document.js',
    '/Users/balakrishna/projects/postman-open-technologies/openapi-componentizer/polaris/node_modules/openapi-type-finder/index.js'
  ]
}

Node.js v19.7.0
```

**Comment by bavulapati** [comment]
@Polaris66 

Overall, I like the approach! 🚀 
Make sure that the code works

**Comment by Polaris66** [comment]
> @Polaris66 I'm running into an error, while using the npm package
> 
> ```
> > ./node_modules/.bin/openapi-type-finder -u https://api.apis.guru/v2/specs/xkcd.com/1.0.0/openapi.yaml -t responses
> node:internal/modules/cjs/loader:1093
>   throw err;
>   ^
> 
> Error: Cannot find module 'express/lib/response'
> Require stack:
> - /Users/balakrishna/projects/postman-open-technologies/openapi-componentizer/polaris/node_modules/openapi-type-finder/document.js
> - /Users/balakrishna/projects/postman-open-technologies/openapi-componentizer/polaris/node_modules/openapi-type-finder/index.js
>     at Module._resolveFilename (node:internal/modules/cjs/loader:1090:15)
>     at Module._load (node:internal/modules/cjs/loader:934:27)
>     at Module.require (node:internal/modules/cjs/loader:1157:19)
>     at require (node:internal/modules/helpers:119:18)
>     at Object.<anonymous> (/Users/balakrishna/projects/postman-open-technologies/openapi-componentizer/polaris/node_modules/openapi-type-finder/document.js:4:18)
>     at Module._compile (node:internal/modules/cjs/loader:1275:14)
>     at Module._extensions..js (node:internal/modules/cjs/loader:1329:10)
>     at Module.load (node:internal/modules/cjs/loader:1133:32)
>     at Module._load (node:internal/modules/cjs/loader:972:12)
>     at Module.require (node:internal/modules/cjs/loader:1157:19) {
>   code: 'MODULE_NOT_FOUND',
>   requireStack: [
>     '/Users/balakrishna/projects/postman-open-technologies/openapi-componentizer/polaris/node_modules/openapi-type-finder/document.js',
>     '/Users/balakrishna/projects/postman-open-technologies/openapi-componentizer/polaris/node_modules/openapi-type-finder/index.js'
>   ]
> }
> 
> Node.js v19.7.0
> ```

Hi, Thanks a lot for your keen observation. This was caused due to a random addition of an extra line in the code. I believe it is fixed now. I have installed and tried the application from another system.

However, if there is still any problem, please inform me if possible.

One thing I have noticed is that the application can only be installed using ```npm install -g``` and not ```npm istall``` and I cannot find a proper explanation for it, can you please help.

**Comment by RaisinTen** [comment]
I can confirm that a global install isn't necessary anymore after the fix. :+1:

I think we need a `.gitignore` file to ignore the `node_modules` directory that gets created when we run `npm ci` in the `package` directory.

**Comment by Polaris66** [comment]
> I can confirm that a global install isn't necessary anymore after the fix. +1
> 
> I think we need a `.gitignore` file to ignore the `node_modules` directory that gets created when we run `npm ci` in the `package` directory.

Hi, does simply adding a .gitignore file with node_modules do the trick? If so, I have done that.

**Comment by RaisinTen** [comment]
Yea, I think you're on the right track. Also, worth noting that `package/node_modules/.package-lock.json` is also being tracked by git despite the fact that the contents of `package/node_modules` isn't. I think deleting `package/node_modules/.package-lock.json` and committing it might fix the issue.

**Comment by Polaris66** [comment]
> Yea, I think you're on the right track. Also, worth noting that `package/node_modules/.package-lock.json` is also being tracked by git despite the fact that the contents of `package/node_modules` isn't. I think deleting `package/node_modules/.package-lock.json` and committing it might fix the issue.

I believe it's fixed now.

File: package/.gitignore [hunk 2]
```diff
@@ -0,0 +1,2 @@
+node_modules
+node_modules/*
```
Comment (bavulapati) [inline package/.gitignore]: both the lines refer to the regex

File: package/README.md [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+                "$ref": "#/components/schemas/comic"
+            }
+        }
+    },
+    "description": "OK"
+}
+```
+***
+
+# Types Done:
```
Comment (bavulapati) [inline package/README.md:40]: I like things being documented but these can have descriptions

**Review by bavulapati — COMMENTED** [review]
You are definitely headed in the right direction

File: package/README.md (hunk 3 above)
Comment (Polaris66) [reply package/README.md:40]: So, should I add a description for each object?

Comment (bavulapati) [reply package/README.md:40]: Description about the checklist - the purpose, etc., but this can wait - not a must

File: package/.gitignore (hunk 2 above)
Comment (Polaris66) [reply package/.gitignore]: Fixed

File: package/README.md (hunk 3 above)
Comment (Polaris66) [reply package/README.md:40]: Fixed

**Comment by Polaris66** [comment]
> You are definitely headed in the right direction

Hi,
I redid a lot of the code to match the specifications given by you in the issues section. Now, all objects that have names shall appear with their names.

Schemas -> [Schema Object]
Responses -> [{Status Code/ "default" : Response Object | Reference Object}]
Parameters -> [Parameter Object | Reference Object]
Examples -> [{name: Example Object | Reference Object}]
Request Bodies -> [ Request Body Object | Reference Object ]
Headers -> [{name: Header Object | Reference Object}]
Links -> [{name: Link Object | Reference Object}]
Security Schemes -> [{name: Link Object | Reference Object}]
Callbacks -> [{name: Callback Object | Reference Object}]
Path Items -> [{path:Path Item Object | Reference Object}]

I think the application should satisfy all needed requirements. Please let me know if I need to make any more changes or is the Qualification Task done.

**Comment by wldeh** [comment]
Hello  @Polaris66! The work you have done here is truly something you should be proud of. One small tip from my end is to consider using https://www.conventionalcommits.org/en/v1.0.0/ to bring more clarity and structure to your work.

**Comment by wldeh** [comment]
One more suggestion I have for you is to use a code formatter like Prettier (https://prettier.io). It can really take your project a long way and improve your code's overall readability and consistency. I don't want to intrude too much, as it seems like @bavulapati  
has everything under control but feel free to contact me if you have any questions!

**Comment by Polaris66** [comment]
> One more suggestion I have for you is to use a code formatter like Prettier (https://prettier.io). It can really take your project a long way and improve your code's overall readability and consistency. I don't want to intrude too much, as it seems like @bavulapati has everything under control but feel free to contact me if you have any questions!

Hi @wldeh, I am extremely grateful for your suggestions and will implement them to the best of my ability. I don't have a lot of experience working on "official" projects and am glad for all inputs.

I also looked through your code and it looks very professionally developed. I am extremely glad for your presence in this community and will be sure to contact you if I need any help.

**Comment by Polaris66** [comment]
> One more suggestion I have for you is to use a code formatter like Prettier (https://prettier.io). It can really take your project a long way and improve your code's overall readability and consistency. I don't want to intrude too much, as it seems like @bavulapati has everything under control but feel free to contact me if you have any questions!

I have used Prettier and used Conventional Commits for the latest commits. It will take me some time getting used to working with Conventional Commits and Autoformatters but I am glad to have begin.

**Pull request closed without merging**